src/btree/bt_handle.c

```diff
@@ -854,12 +854,12 @@ __btree_get_last_recno(WT_SESSION_IMPL *session)
     }
 
     /*
-     * The endpoint for append is global; read the last page with global visibility to make sure
-     * that if the end of the tree is truncated we don't start appending in the truncated space
-     * unless the truncation has become globally visible. (Note that this path does not examine the
-     * visibility of individual data items; it only checks whether whole pages are deleted.)
+     * The endpoint for append is global; read the last page with global visibility (even if it's
+     * deleted) to make sure that if the end of the tree is truncated, the tree walk finds the
+     * correct page. (Note that this path does not examine the visibility of individual data items;
+     * it only checks whether whole pages are deleted.)
      */
-    flags = WT_READ_PREV | WT_READ_VISIBLE_ALL;
+    flags = WT_READ_PREV | WT_READ_VISIBLE_ALL | WT_READ_SEE_DELETED;
 
     next_walk = NULL;
     WT_RET(__wt_tree_walk(session, &next_walk, flags));
```

Comment:
After understanding the problem, do we really need this tree walk? Why can't it also directly go to the end of the last page and calculate the last recno similar to the search? 

This may be a significant change from the current one, I want to make sure that both the code flows are using the same approach so that this kind of bug doesn't repeat if we mistakenly introduce a bug with the flag approach.

Reply:
We could do it the same way as search, but I would personally be inclined to go the other way around - why doesn't search do a tree walk?

If there are good reasons we don't use a tree walk for search, then I think we should factor out that search code before trying to use it in other places. That seems like a bigger task than is suitable for this ticket, but I'm flexible there.

Reply:
It may be due to the performance reasons that the search doesn't use the tree walk. I am fine to do the change of aligning the search and last recno into the same code flow as part of a seperate ticket.

Reply:
Made WT-11653 to track this.